File: src/data/shipsystems/scripts/aEP_DroneDash.java
package data.shipsystems.scripts;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.combat.*;
import com.fs.starfarer.api.impl.combat.BaseShipSystemScript;
import combat.plugin.aEP_BuffEffect;
import combat.impl.aEP_Buff;
import combat.util.aEP_Tool;
import org.lazywizard.lazylib.MathUtils;
import org.lwjgl.util.vector.Vector2f;

import java.awt.*;

public class aEP_DroneDash extends BaseShipSystemScript
{
  static final float MAX_SPEED_BONUS = 200f;
  static final float TURN_RATE_BONUS = 250f;
  static final float ROTATE_SPEED = 20f;

  static final float END_BUFF_TIME = 1f;
  static final float END_TURN_RATE_BONUS = 300f;
  static final float END_TURN_ACC_BONUS = 50f;

  static final float DAMAGE_TAKEN = 0.25f;

  @Override
  public void apply(MutableShipStatsAPI stats, String id, State state, float effectLevel) {
    ShipAPI ship = (ShipAPI) stats.getEntity();
    if (ship == null) return;

    float amount = Global.getCombatEngine().getElapsedInLastFrame() * stats.getTimeMult().getModifiedValue();

    stats.getMaxSpeed().modifyFlat(id, Math.max(effectLevel, 0.5f) * MAX_SPEED_BONUS);
    stats.getAcceleration().modifyFlat(id, MAX_SPEED_BONUS * 2f);
    stats.getDeceleration().modifyMult(id, 0);

    stats.getEngineDamageTakenMult().modifyMult(id, DAMAGE_TAKEN);
    stats.getArmorDamageTakenMult().modifyMult(id,DAMAGE_TAKEN);
    stats.getHullDamageTakenMult().modifyMult(id,DAMAGE_TAKEN);

    ship.blockCommandForOneFrame(ShipCommand.ACCELERATE_BACKWARDS);
    ship.blockCommandForOneFrame(ShipCommand.DECELERATE);
    ship.blockCommandForOneFrame(ShipCommand.STRAFE_RIGHT);
    ship.blockCommandForOneFrame(ShipCommand.STRAFE_LEFT);
    ship.giveCommand(ShipCommand.ACCELERATE, null, 0);

    Vector2f angleAndSpeed = aEP_Tool.Util.velocity2Speed(ship.getVelocity());
    angleAndSpeed.y += ship.getAcceleration() * amount;
    ship.getVelocity().set(aEP_Tool.Util.speed2Velocity(angleAndSpeed.x, angleAndSpeed.y));


    //下面只在完全激活时跑一次
    if (effectLevel < 1) return;
    CombatEngineAPI engine = Global.getCombatEngine();
    //速度归0
    ship.getVelocity().set(new Vector2f(0, 0));
    //加烟
    int num = 16;
    for (int i = 0; i < num; i++) {
      engine.addNebulaSmokeParticle(MathUtils.getRandomPointInCircle(ship.getLocation(), ship.getCollisionRadius()),
        aEP_Tool.Util.speed2Velocity(ship.getFacing() - 180f, MathUtils.getRandomNumberInRange(50, 100)),

        MathUtils.getRandomNumberInRange(20, 60),
        2f,
        0.25f,
        0.5f,
        1,
        new Color(255, 250, 250, 125));
    }
    //加残影
    Vector2f vel = aEP_Tool.Util.speed2Velocity(ship.getFacing(), -100);
    ship.addAfterimage(new Color(255, 155, 155, 250),
      0, 0,
      vel.x, vel.y,
      5,
      0,
      0.5f,
      0.5f,
      true,
      false,
      true);
    vel = aEP_Tool.Util.speed2Velocity(ship.getFacing(), -200);
    ship.addAfterimage(new Color(255, 155, 155, 250),
      0, 0,
      vel.x, vel.y,
      5,
      0,
      0.5f,
      0.5f,
      true,
      false,
      true);
    vel = aEP_Tool.Util.speed2Velocity(ship.getFacing(), -300);
    ship.addAfterimage(new Color(255, 155, 155, 250),
      0, 0,
      vel.x, vel.y,
      5,
      0,
      0.5f,
      0.5f,
      true,
      false,
      true);
    MissileAPI m = null;

  }

  @Override
  public void unapply(MutableShipStatsAPI stats, String id) {
    stats.getMaxSpeed().unmodify(id);
    stats.getAcceleration().unmodify(id);
    stats.getDeceleration().unmodify(id);
    stats.getEngineDamageTakenMult().unmodify(id);
    stats.getArmorDamageTakenMult().unmodify(id);
    stats.getHullDamageTakenMult().unmodify(id);
    ShipAPI ship = (ShipAPI) stats.getEntity();
    if (ship != null)
      aEP_BuffEffect.addThisBuff(ship, new aEP_ExtraTurnRate(ship));
  }

  class aEP_ExtraTurnRate extends aEP_Buff
  {
    public aEP_ExtraTurnRate(ShipAPI ship) {
      setStackNum(1f);
      setMaxStack(1f);
      setEntity(ship);
      setRenew(true);
      setLifeTime(END_BUFF_TIME);
      setBuffType("aEP_ExtraTurnRate");
    }

    @Override
    public void play() {
      ShipAPI ship = (ShipAPI) getEntity();
      ship.getMutableStats().getMaxTurnRate().modifyPercent(getBuffType(), END_TURN_RATE_BONUS);
      ship.getMutableStats().getTurnAcceleration().modifyPercent(getBuffType(), END_TURN_RATE_BONUS * 2);
      ship.getMutableStats().getAcceleration().modifyPercent(getBuffType(), END_TURN_ACC_BONUS);
      ship.getMutableStats().getDeceleration().modifyPercent(getBuffType(), END_TURN_ACC_BONUS);
      ship.setJitterUnder(getBuffType(), new Color(255, 155, 155, 255), 1, 18, 1);

    }

    @Override
    public void readyToEnd() {
      ShipAPI ship = (ShipAPI) getEntity();
      ship.getMutableStats().getMaxTurnRate().unmodify(getBuffType());
      ship.getMutableStats().getTurnAcceleration().unmodify(getBuffType());
      ship.getMutableStats().getAcceleration().unmodify(getBuffType());
      ship.getMutableStats().getDeceleration().unmodify(getBuffType());
    }
  }
}

File: src/data/shipsystems/scripts/aEP_DroneDash.kt
package data.shipsystems.scripts

import com.fs.starfarer.api.Global
import combat.util.aEP_Tool.Util.velocity2Speed
import combat.util.aEP_Tool.Util.speed2Velocity
import combat.plugin.aEP_BuffEffect.Companion.addThisBuff
import com.fs.starfarer.api.impl.combat.BaseShipSystemScript
import com.fs.starfarer.api.combat.MutableShipStatsAPI
import com.fs.starfarer.api.plugins.ShipSystemStatsScript
import com.fs.starfarer.api.combat.ShipAPI
import data.shipsystems.scripts.aEP_DroneDash
import com.fs.starfarer.api.combat.ShipCommand
import org.lwjgl.util.vector.Vector2f
import combat.util.aEP_Tool
import com.fs.starfarer.api.combat.CombatEngineAPI
import org.lazywizard.lazylib.MathUtils
import com.fs.starfarer.api.combat.MissileAPI
import combat.plugin.aEP_BuffEffect
import data.shipsystems.scripts.aEP_DroneDash.aEP_ExtraTurnRate
import combat.impl.aEP_Buff
import java.awt.Color

class aEP_DroneDash : BaseShipSystemScript() {
  companion object {
    const val MAX_SPEED_BONUS = 200f
    const val TURN_RATE_BONUS = 250f
    const val ROTATE_SPEED = 20f
    const val END_BUFF_TIME = 1f
    const val END_TURN_RATE_BONUS = 300f
    const val END_TURN_ACC_BONUS = 50f
    const val DAMAGE_TAKEN = 0.33f

    val AFTER_MERGE_COLOR = Color(255, 155, 155, 250)
    val SMOKE_MERGE_COLOR = Color(255, 250, 250, 125)
  }

  override fun apply(stats: MutableShipStatsAPI, id: String, state: ShipSystemStatsScript.State, effectLevel: Float) {
    //stats的entity有可能为null
    val ship = (stats?.entity?: return)as ShipAPI

    val amount = Global.getCombatEngine().elapsedInLastFrame * stats.timeMult.modifiedValue
    stats.maxSpeed.modifyFlat(id, Math.max(effectLevel, 0.5f) * MAX_SPEED_BONUS)
    stats.acceleration.modifyFlat(id, MAX_SPEED_BONUS * 2f)
    stats.deceleration.modifyMult(id, 0f)
    stats.engineDamageTakenMult.modifyMult(id, DAMAGE_TAKEN)
    stats.armorDamageTakenMult.modifyMult(id, DAMAGE_TAKEN)
    stats.hullDamageTakenMult.modifyMult(id, DAMAGE_TAKEN)
    ship.blockCommandForOneFrame(ShipCommand.ACCELERATE_BACKWARDS)
    ship.blockCommandForOneFrame(ShipCommand.DECELERATE)
    ship.blockCommandForOneFrame(ShipCommand.STRAFE_RIGHT)
    ship.blockCommandForOneFrame(ShipCommand.STRAFE_LEFT)
    ship.giveCommand(ShipCommand.ACCELERATE, null, 0)
    val angleAndSpeed = velocity2Speed(ship.velocity)
    angleAndSpeed.y += ship.acceleration * amount
    ship.velocity.set(speed2Velocity(angleAndSpeed.x, angleAndSpeed.y))


    //下面只在完全激活时跑一次
    if (effectLevel < 1) return
    val engine = Global.getCombatEngine()
    //速度归0
    ship.velocity.set(Vector2f(0f, 0f))
    //加烟
    val num = 16
    for (i in 0 until num) {
      engine.addNebulaSmokeParticle(
        MathUtils.getRandomPointInCircle(ship.location, ship.collisionRadius),
        speed2Velocity(ship.facing - 180f, MathUtils.getRandomNumberInRange(50, 100).toFloat()),
        MathUtils.getRandomNumberInRange(20, 60).toFloat(),
        2f,
        0.25f,
        0.5f, 1f,
        SMOKE_MERGE_COLOR
      )
    }
    //加残影
    var vel = speed2Velocity(ship.facing, -100f)
    ship.addAfterimage(
      AFTER_MERGE_COLOR, 0f, 0f,
      vel.x, vel.y, 5f, 0f,
      0.5f,
      0.5f,
      true,
      false,
      true
    )
    vel = speed2Velocity(ship.facing, -200f)
    ship.addAfterimage(
      AFTER_MERGE_COLOR, 0f, 0f,
      vel.x, vel.y, 5f, 0f,
      0.5f,
      0.5f,
      true,
      false,
      true
    )
    vel = speed2Velocity(ship.facing, -300f)
    ship.addAfterimage(
      AFTER_MERGE_COLOR, 0f, 0f,
      vel.x, vel.y, 5f, 0f,
      0.5f,
      0.5f,
      true,
      false,
      true
    )
    val m: MissileAPI? = null
  }

  override fun unapply(stats: MutableShipStatsAPI, id: String) {
    //stats的entity有可能为null
    val ship = (stats?.entity?: return)as ShipAPI

    stats.maxSpeed.unmodify(id)
    stats.acceleration.unmodify(id)
    stats.deceleration.unmodify(id)
    stats.engineDamageTakenMult.unmodify(id)
    stats.armorDamageTakenMult.unmodify(id)
    stats.hullDamageTakenMult.unmodify(id)
    addThisBuff(ship, aEP_ExtraTurnRate(ship))
  }

  internal inner class aEP_ExtraTurnRate(ship: ShipAPI) : aEP_Buff() {
    override fun play() {
      val ship = entity as ShipAPI
      ship.mutableStats.maxTurnRate.modifyPercent(buffType, END_TURN_RATE_BONUS)
      ship.mutableStats.turnAcceleration.modifyPercent(buffType, END_TURN_RATE_BONUS * 2)
      ship.mutableStats.acceleration.modifyPercent(buffType, END_TURN_ACC_BONUS)
      ship.mutableStats.deceleration.modifyPercent(buffType, END_TURN_ACC_BONUS)
      ship.setJitterUnder(buffType, Color(255, 155, 155, 255), 1f, 18, 1f)
    }

    override fun readyToEnd() {
      val ship = entity as ShipAPI
      ship.mutableStats.maxTurnRate.unmodify(buffType)
      ship.mutableStats.turnAcceleration.unmodify(buffType)
      ship.mutableStats.acceleration.unmodify(buffType)
      ship.mutableStats.deceleration.unmodify(buffType)
    }

    init {
      stackNum = 1f
      maxStack = 1f
      entity = ship
      isRenew = true
      lifeTime = END_BUFF_TIME
      buffType = "aEP_ExtraTurnRate"
    }
  }

}